### PR TITLE
Transplant MPAS code base to E3SM

### PR DESCRIPTION
**Pull request by mark-petersen**
Move MPAS from a submodule to code within E3SM. All functionality in E3SM is identical to before. Subtree commands were used to keep the commit history for each MPAS component.

MPAS stand-alone may still be compiled in the `components/mpas-framework` directory.

Diagram of revised directory structure
image

See line-by-line instructions for how this PR was created [in the comment below](https://github.com/E3SM-Project/E3SM/pull/4214#issuecomment-818208881)

[Instructions on how to convert PRs from MPAS to E3SM](https://github.com/E3SM-Project/E3SM/pull/4214#issuecomment-832007454)

[BFB]

### CODE REVIEW

**Comment by mark-petersen** [comment]
Diagram of directory structure
image

**Comment by mark-petersen** [comment]
How I created this branch:

**MPAS-Model Repository**
**1. Create branch of e3sm/develop for move**
```
git checkout -b e3sm_develop_for_mpas_move origin/e3sm/develop
git submodule update --init --recursive
```

**2. Remove submodules**
```
export DIR=BGC
git submodule deinit -f -- src/core_ocean/$DIR
rm -rf .git/modules/src/core_ocean/$DIR
git rm -f src/core_ocean/$DIR
git commit -m "Remove submodule $DIR"
```

Repeat step for:
export DIR=cvmix

**3. Split off cores**
```
export CORE=ocean
git subtree split --prefix=src/core_$CORE --branch=subtree_$CORE
git push mark-petersen subtree_$CORE 
git rm -r src/core_$CORE
git commit -am "Remove $CORE core for subtree transplant to E3SM"
```

Repeat step for:
export CORE=seaice
export CORE=landice

**4. Remove atmospheric cores, push branch**
```
git rm -r src/core_atmosphere
git rm -r src/core_init_atmosphere
git commit -am "Remove atmospheric core for subtree transplant to E3SM"
git push mark-petersen e3sm_develop_for_mpas_move
```

**E3SM Repository**
**1. Create branch of master for move**
```
git checkout -b mpas_move origin/master
git submodule update --init --recursive
```

**2. Remove mpas-source submodules**
```
export DIR=components/mpas-source
git submodule deinit -f -- $DIR
rm -rf .git/modules/$DIR
git rm -f $DIR
git commit -m "Remove components/mpas-source submodule"
```

**3. Add cores back in**
```
export CORE=ocean
git subtree add --prefix=components/mpas-${CORE}/src git@github.com:mark-petersen/MPAS-Model.git \
    subtree_$CORE
```

Repeat step for:
export CORE=seaice
export CORE=landice

**4. Add MPAS framework back in**
```
git subtree add --prefix=components/mpas-framework git@github.com:mark-petersen/MPAS-Model.git \
   e3sm_develop_for_mpas_move
```

**5. Add ocean submodules back in**
```
git submodule add https://github.com/CVMix/CVMix-src.git components/mpas-ocean/src/cvmix
git commit -m "Add cvmix submodule to ocean"

git submodule add https://github.com/E3SM-Project/Ocean-BGC.git components/mpas-ocean/src/BGC
git commit -m "Add BGC submodule to ocean"
```

**Review by jgfouca — APPROVED** [review]

**Comment by jgfouca** [comment]
@mark-petersen , these changes look fine. I was just curious as to what was the motivation for this change?

**Comment by mark-petersen** [comment]
@jgfouca here are notes we had from a meeting Feb 9. [Also on Confluence](https://acme-climate.atlassian.net/wiki/spaces/EIDMG/pages/2280194361/MPAS+E3SM+repository+workflow)

**Issues with current workflow**

1. Coupled ocean and seaice PRs and development is complicated by multiple branches and repos
2. Required to learn two very different (and complicated) workflows for building and developing MPAS vs. E3SM

A lengthier set of requirements was also discussed on that page. These include simplicity of workflow, keeping the MPAS commit history, ability to build MPAS-stand alone, etc. We went through that list and a number of possible solutions, and settled on transplanting the MPAS code into E3SM using the subtree functionality. 

The workflow for MPAS ocean and seaice development moving forward will be to make code changes and PRs directly into E3SM-Project/E3SM.

**Comment by xylar** [comment]
@mark-petersen and @matthewhoffman, I'm seeing:
```
$ cd components/
$ ls -1
...
mpas-albany-landice
...
mpas-landice
...
```
and then
```
$ ls mpas-albany-landice
bld  cime_config  driver  tools
$ ls mpas-landice
src
``
This raised a few questions for me about MALI and this plan.

First, I don't recall how we decided to handle MALI and maybe that was never made clear.  Did we decide it would be handled like ocean and sea ice?  If so, that should probably be part of the changes mentioned in the diagram above. 

Second, shouldn't the `src` directory for MALI be in `mpas-albandy-landice`, not `mpas-landice` if it's going to get put somewhere?

Third, (and maybe this is more for @matthewhoffman), how is MALI going to be updated in E3SM with this new structure?  With subtrees?

**Comment by xylar** [comment]
Other than my questions above, things look great to me.  I tried making a new submodule in the `compass` package and building from this branch, and everything worked great including bit-for-bit results in the ocean nightly test suite.

**Comment by matthewhoffman** [comment]
@mark-petersen , we had a MALI team meeting today to discuss future MALI development.  Our tentative consensus was to follow the lead of the ocean and seaice cores and also move to MALI to the E3SM repo.  As such, I would like to coordinate with you to update this PR to also include moving MALI.

cc: @stephenprice , @mperego , @trhille

**Comment by mark-petersen** [comment]
@matthewhoffman  I'm happy to work with you on that. Here is what I propose:
1. @mark-petersen  merges `landice/develop` into `e3sm/develop` locally.
2. @mark-petersen  recreates this PR with updated mpas-albany-landice
3. @matthewhoffman tests this PR for both standalone and E3SM functionality, revise as needed.
4. @matthewhoffman makes E3SM PR for merge of `landice/develop` to `e3sm/develop`
5. This PR is ready to merge.

Let me know if I should proceed with (1), or if you want me to wait for you to merge more items in.

@xylar, that naming with `mpas-landice` was my error in the commands. Old habits die hard! I will change to `mpas-albany-landice` on this next round. Thanks for catching it.

**Comment by mark-petersen** [comment]
@jgfouca if you look at this commit,
https://github.com/E3SM-Project/E3SM/pull/4214/commits/3520bc6735d0bc00323d7e19e49d02122ecbe8be
you will see that I put three links inside the `mpas-framework` directory (previously `mpas-source`), to the code for each core. I did this so that the make and cmake files could all remain the same for this PR. I think it makes more sense for the cmake files to point directly to the code directories. I tried it myself, but was not successful.

What are your thoughts on this? Are the links OK for this PR, or should we change the cmake files here?

The directory image at https://github.com/E3SM-Project/E3SM/pull/4214#issuecomment-817845238 might be useful:

**Comment by mark-petersen** [comment]
Now passes 
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_gnu
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
PET_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu
```

**Comment by mark-petersen** [comment]
@rljacob this is a major change in the configuration of the MPAS files within E3SM. Please review.

We also liked your idea for another organization to hold PRs for longer-term discussions and development, since PRs to `E3SM-Project/E3SM` are meant for quick merging. Each organization only allows one fork, which is why we need a new organization.  For example, `E3SM-Draft/E3SM`. When a PR on that organization is ready to merge, a new PR would be made to `E3SM-Project/E3SM`. Let us know your thoughts on that as well, and how to proceed.

**Comment by jgfouca** [comment]
@mark-petersen , I think links are fine for now.

**Review by rljacob — APPROVED** [review]

**Comment by mark-petersen** [comment]
This is how you move a current MPAS PR to be an E3SM PR. This example is for the ocean core, but other cores substitute your name for ocean. 

**MPAS-Model Repository**
**1. Rebase PR on head of e3sm/develop, commit `83b74e7ec`**

**2. Split out core**
```
git subtree split --prefix=src/core_ocean --branch=subtree_move_PR
git push YourMPASFork subtree_move_PR
```
The commit hashes are always identical to previous splits. That is great!

**E3SM Repository**
**1. Add updated code with subtree merge and pull**
This prepares the local E3SM branch:
```
git reset --hard origin/master
git submodule update --init --recursive
git subtree merge --prefix=components/mpas-ocean/src HEAD
```
This brings the new commits over:
```
git subtree pull --prefix=components/mpas-ocean/src \
   git@github.com:YourMPASFork/MPAS-Model.git subtree_move_PR
```
Now make a new pull request to E3SM.

**Comment by mark-petersen** [comment]
In MPAS stand-alone, compiles with ocean, seaice, and landice cores. Passes ocean nightly regression suite.

In E3SM, passes:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_gnu
PET_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu
```

File: .gitmodules [hunk 1]
```diff
@@ -43,3 +40,12 @@
 	url = git@github.com:NVlabs/cub.git
 [submodule "Submod"]
 	branch = rljacob/e3smconfig/elm-name
+[submodule "components/mpas-ocean/src/cvmix"]
+	path = components/mpas-ocean/src/cvmix
+	url = https://github.com/CVMix/CVMix-src.git
```
Comment (matthewhoffman) [inline .gitmodules:45]: @mark-petersen , the three new ocean code submodules have been added using the https protocol, but all other E3SM submodules use the ssh protocol.  Is that a mistake?

File: components/mpas-framework/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
+MPAS-v7.0
```
Comment (matthewhoffman) [inline components/mpas-framework/README.md:1]: This file should be rewritten to describe the code that's actually in E3SM and how it's organized.  That could happen in an immediate follow-on PR.

**Comment by xylar** [comment]
@philipwjones had a preference for HTTPS if I recall correctly.  You don't need a GitHub account or SSH keys to use it.  But I also have seen messages on GitHub that they don't plan to continue supporting HTTPS for much longer so I don't know if that's really a good consideration at this point.  I'll send a link once I find it.

**Comment by mattdturner** [comment]
This passes my usual Compy tests:
```
PASS:
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_intel
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_intel
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_intel
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_intel
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi.C
  - PASS SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi BASELINE /compyfs/e3sm_baselines/pgi/master: 
SMS_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PET_Ln3_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.compy_intel
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.compy_pgi
```

**Review by mattdturner — APPROVED** [review]

File: .gitmodules (hunk 1 above)
Comment (mark-petersen) [reply .gitmodules]: OK, I changed it to git@github now and it all works. Thanks.

**Comment by rljacob** [comment]
We've always used ssh but that dates to when the source was closed.   We should stick to one protocol for all the submodules within E3SM.

**Comment by mark-petersen** [comment]
OK, that sounds reasonable. I changed the three MPAS submodules to ssh on the last commit.

**Comment by philipwjones** [comment]
@mark-petersen On summit, I ran

SMS.T62_oQU240.GMPAS-IAF.summit_ibm
SMS.T62_oQU240.GMPAS-IAF.summit_pgi

and both passed.

**Comment by mark-petersen** [comment]
Recreated this branch on today's master, with today's MPAS `ocean/develop`. Passes:
```
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_gnu
PET_Ln9.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-knl_gnu
```

**Comment by mattdturner** [comment]
```
PASS:
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi.C
  - PASS SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi BASELINE /compyfs/e3sm_baselines/pgi/master:
SMS_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PET_Ln3_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.compy_pgi
```

**Comment by jonbob** [comment]
test merge passed sanity testing

**Comment by jonbob** [comment]
merged to next -- @mark-petersen , you made this very easy! Thanks

**Comment by mark-petersen** [comment]
Thanks. Fingers crossed, it will be this easy on the test boards tomorrow.

**Comment by jonbob** [comment]
merged to master

**Pull request merged**